### PR TITLE
dialects: (llvm) Implement custom format for FuncOp

### PR DESCRIPTION
**Pull request by sueszli**
Allows not just generic format:

```mlir
"llvm.func"() <{
  sym_name = "binops",
  function_type = !llvm.func<void (i32, i32, f32, f32)>,
  CConv = #llvm.cconv<ccc>,
  linkage = #llvm.linkage<external>,
  visibility_ = 0 : i64
}> ({
^bb0(%arg0 : i32, %arg1 : i32, %arg2 : f32, %arg3 : f32):
  %0 = llvm.add %arg0, %arg1 : i32
  // ... ops ...
  "llvm.return"() : () -> ()
}) : () -> ()

```

But also a prettier custom format:

```mlir
llvm.func @binops(%arg0: i32, %arg1: i32, %arg2: f32, %arg3: f32) {
  %0 = llvm.add %arg0, %arg1 : i32
  %1 = llvm.fadd %arg2, %arg3 : f32
  // ... ops ...
  llvm.return
}

### CODE REVIEW

File: xdsl/ir/core.py [hunk 1]
```diff
@@ -1396,6 +1396,10 @@ def verify_(self) -> None:
     def parse(cls, parser: Parser) -> Self:
         parser.raise_error(f"Operation {cls.name} does not have a custom format.")
 
+    # mirrors https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/TableGen/AttrOrTypeDef.h#L197
+    def should_use_custom_format(self) -> bool:
+        return True
```
Comment (superlopuh) [inline xdsl/ir/core.py]: ? Why is this necessary?

Comment (superlopuh) [reply xdsl/ir/core.py]: I don't see how this corresponds to the MLIR API

File: xdsl/printer.py [hunk 2]
```diff
@@ -607,8 +607,12 @@ def print_op(self, op: Operation) -> None:
...
             self.print_string(f'"{op.op_name.data}"')
         # If we print with the generic format, or the operation does not have a custom
-        # format
-        elif self.print_generic_format or Operation.print is type(op).print:
+        # format, or the operation signals it wants generic format
+        elif (
+            self.print_generic_format
+            or Operation.print is type(op).print
+            or not op.should_use_custom_format()
+        ):
```
Comment (superlopuh) [inline xdsl/printer.py]: ops can always fall back to the default implementation of `printer.print_op_with_default_format(self)`

File: xdsl/dialects/utils/format.py [hunk 3]
```diff
@@ -181,10 +180,16 @@ def parse_for_op_like(
...
+class FunctionTypeLike(Protocol):
+    @property
+    def outputs(self) -> Any: ...
+
+
 def print_func_op_like(
     printer: Printer,
     sym_name: StringAttr,
-    function_type: FunctionType,
+    function_type: Attribute,
```
Comment (superlopuh) [inline xdsl/dialects/utils/format.py]: We have at least one place (but maybe more) where we can pass either a FunctionType or a tuple of sequences of types for the inputs and outputs, maybe we can do that here?

File: tests/filecheck/dialects/llvm/func.mlir [hunk 4]
```diff
@@ -46,4 +46,35 @@
...
-// CHECK-NEXT:  }
+
+// Checks support for custom format
+
+llvm.func @external_func(i64)
+
+// CHECK: llvm.func @external_func(i64)
+
+llvm.func @void_func(%arg0: i64) {
+  "llvm.return"() : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/llvm/func.mlir]: this also looks a little sad, this will take about 30 seconds to add a custom syntax for

File: xdsl/dialects/llvm.py [hunk 5]
```diff
@@ -1687,6 +1699,157 @@ def __init__(
...
+            "available_externally",
+            "linkonce",
+            "weak",
+            "common",
+            "appending",
+            "extern_weak",
+            "linkonce_odr",
+            "weak_odr",
+            "external",
+        ]
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: let's take this out so that it's not redefined on every parse? Similarly below for collections of strings

File: xdsl/dialects/llvm.py [hunk 6]
```diff
@@ -1687,6 +1699,157 @@ def __init__(
...
+        linkage = None
+        for l in linkage_options:
+            if parser.parse_optional_keyword(l):
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:1729]: is this how the MLIR code works?

File: xdsl/dialects/llvm.py [hunk 7]
```diff
@@ -1687,6 +1699,157 @@ def __init__(
...
+        if len(return_types) == 0:
+            return_type = LLVMVoidType()
+        elif len(return_types) == 1:
+            return_type = return_types[0]
+        else:
+            parser.raise_error(
+                "llvm.func only supports a single return type (or void)",
+                parser.pos,
+                parser.pos,
+            )
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: Does MLIR raise a parse error or verification error? I would by default just let verification deal with it, but would make sense to follow MLIR logic.

**Comment by sueszli** [comment]
Note to self: Increase coverage

File: xdsl/dialects/llvm.py (hunk 6 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py:1729]: By this I mean, does it try for each linkage option, or parse an optional keyword and then introspect the keyword?

File: tests/filecheck/dialects/llvm/func.mlir (hunk 4 above)
Comment (sueszli) [reply tests/filecheck/dialects/llvm/func.mlir]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/changes/17bad16d42239dc04969d563ea04530281fc2aad

File: xdsl/dialects/utils/format.py (hunk 3 above)
Comment (sueszli) [reply xdsl/dialects/utils/format.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/changes/38ebd2937ebe93c0aac3b2f2a597540ed727aaa4

File: xdsl/ir/core.py (hunk 1 above)
Comment (sueszli) [reply xdsl/ir/core.py]: Sorry, the comment on MLIR is wrong: https://github.com/xdslproject/xdsl/pull/5588/changes/1964563caa7f9d1fd58b381b7421773a44bbec1c

File: xdsl/dialects/utils/format.py [hunk 8]
```diff
@@ -184,14 +183,20 @@ def parse_for_op_like(
 def print_func_op_like(
     printer: Printer,
     sym_name: StringAttr,
-    function_type: FunctionType,
+    function_type: Attribute | tuple[Sequence[Attribute], Sequence[Attribute]],
```
Comment (superlopuh) [inline xdsl/dialects/utils/format.py]: How about this?
```suggestion
    function_type: FunctionType | tuple[Sequence[Attribute], Sequence[Attribute]],
```

Comment (superlopuh) [reply xdsl/dialects/utils/format.py]: Actually I think I'm being silly, it's basically free here to just create a FunctionType from an LLVMFunctionType, let's just do that and call it a day

File: xdsl/dialects/llvm.py [hunk 9]
```diff
@@ -236,6 +241,12 @@ def __init__(
     def is_variadic(self) -> bool:
         return isinstance(self.variadic, UnitAttr)
 
+    @property
+    def outputs(self) -> Sequence[Attribute]:
+        if isinstance(self.output, LLVMVoidType):
+            return []
+        return [self.output]
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: let's not do this, and inline this code when calling `print_func_op_like`?

File: xdsl/printer.py (hunk 2 above)
Comment (sueszli) [reply xdsl/printer.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/9f8c676d80a4fe164dc508826dc09df89a8844ca

File: xdsl/ir/core.py (hunk 1 above)
Comment (sueszli) [reply xdsl/ir/core.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/9f8c676d80a4fe164dc508826dc09df89a8844ca

File: xdsl/dialects/llvm.py (hunk 9 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/a82bcb15478761512a9939d38d649da822080597

File: xdsl/dialects/utils/format.py (hunk 8 above)
Comment (sueszli) [reply xdsl/dialects/utils/format.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/b1b2580a598e285f7c12499f052cef3a2678a852

Comment (sueszli) [reply xdsl/dialects/utils/format.py]: > just create a FunctionType from an LLVMFunctionType

Sorry may I please ask you to elaborate?

File: xdsl/dialects/llvm.py (hunk 6 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1729]: Equivalent bits:

- https://github.com/llvm/llvm-project/blob/60556a2ec228133638083b29f0cc5609a04b4dfc/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp#L3076
- https://github.com/llvm/llvm-project/blob/60556a2ec228133638083b29f0cc5609a04b4dfc/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp#L92

Comment (sueszli) [reply xdsl/dialects/llvm.py:1729]: > does it try for each linkage option, or parse an optional keyword and then introspect the keyword?

Thanks for clarifying.

If I understood this correctly, the MLIR implementation tries each linkage option one by one and one for every possible value in the Linkage enum. It doesn't do keyword introspection.

File: xdsl/dialects/llvm.py (hunk 7 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: See: https://github.com/llvm/llvm-project/blob/60556a2ec228133638083b29f0cc5609a04b4dfc/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp#L3037

MLIR also treats it as a parse error.

**Comment by sueszli** [comment]
Todo:

- increasing coverage
- figuring out the right `function_type`

File: xdsl/dialects/llvm.py [hunk 10]
```diff
@@ -1687,6 +1707,119 @@ def __init__(
...
+        return FuncOp(
+            name,
+            function_type,
+            LinkageAttr(linkage) if linkage else LinkageAttr("external"),
+            CallingConventionAttr(cconv) if cconv else CallingConventionAttr("ccc"),
+            visibility=0,
+            sym_visibility=visibility,
+            body=region,
+            other_props={
+                **(extra_attrs.data if extra_attrs else {}),
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: Shouldn't this be set as attributes? Can you please add a test for this?

File: xdsl/dialects/llvm.py [hunk 11]
```diff
@@ -1687,6 +1707,119 @@ def __init__(
...
+            printer.print_string(")")
+            if not isinstance(self.function_type.output, LLVMVoidType):
+                printer.print_string(" -> ")
+                printer.print_attribute(self.function_type.output)
+            printer.print_op_attributes(
+                attrs,
+                reserved_attr_names=FUNC_OP_RESERVED_ATTR_NAMES,
+                print_keyword=True,
+            )
+            return
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: Why is this necessary? does `print_func_op_like` not satisfy the format for the llvm func? in `func.func` we use the helper both for declarations and definitions.

File: xdsl/dialects/llvm.py [hunk 12]
```diff
@@ -1687,6 +1707,119 @@ def __init__(
...
+        outputs = (
+            []
+            if isinstance(self.function_type.output, LLVMVoidType)
+            else [self.function_type.output]
+        )
+
+        print_func_op_like(
+            printer,
+            self.sym_name,
+            (self.function_type.inputs.data, outputs),
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: This is what I mean, this way the helper doesn't have to change, it's a minor overhead
```suggestion
            FunctionType(self.function_type.inputs.data, outputs),
```

File: xdsl/dialects/utils/format.py (hunk 8 above)
Comment (superlopuh) [reply xdsl/dialects/utils/format.py]: Sorry for lack of clarity on my part, elaborated below in an edit suggestion

Comment (sueszli) [reply xdsl/dialects/utils/format.py]: Thank you!

File: xdsl/dialects/llvm.py (hunk 10 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/6f2b588631cab28fa46cd78f6a20f77b801e209f (I believe)?

File: xdsl/dialects/llvm.py (hunk 12 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Clever, thank you!

Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/ae4d896ec140796b095baa5504b87dcad10c8356

File: xdsl/dialects/llvm.py (hunk 11 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/879c20a7e9ef61aa2264dd96309260cea518149a

File: xdsl/dialects/llvm.py [hunk 13]
```diff
@@ -1723,7 +1842,7 @@ def parse_value(cls, parser: Parser) -> Attribute:
         attr = parser.parse_optional_attribute()
         if attr:
             return attr
-        return IntegerAttr(parser.parse_integer(), 64)
+        return parser.parse_attribute()
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: seems like this was meant for a different PR?

Comment (sueszli) [reply xdsl/dialects/llvm.py]: Whoops - leftover from a hack I tried to reach 100% coverage. Thanks for pointing out.

Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/7f17ee43737c1a9d7a7c6a9f07da2cb4e4b8d38c

File: tests/dialects/test_llvm.py
Comment (superlopuh) [inline tests/dialects/test_llvm.py:1]: were these changes intended to be included in this PR?

File: xdsl/dialects/llvm.py [hunk 14]
```diff
@@ -1696,6 +1813,8 @@ class ReturnOp(IRDLOperation):
 
     name = "llvm.return"
 
+    assembly_format = "($arg^ `:` type($arg))? attr-dict"
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:1891]: I'm now thinking that this PR is dragging on a bit, let's merge this change in its own tiny PR just to reduce the load here a bit?

File: tests/dialects/test_llvm.py
Comment (sueszli) [reply tests/dialects/test_llvm.py:1]: They could be a seperate PR where I just try to maximize coverage. I agree. This is getting very confusing.

File: xdsl/dialects/llvm.py (hunk 14 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1891]: Good idea.

File: tests/filecheck/dialects/llvm/global.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/llvm/global.mlir:1]: this seems like a spurious change?

File: tests/dialects/test_llvm.py
Comment (superlopuh) [reply tests/dialects/test_llvm.py:1]: Hmm, I'm wondering what you mean. Do you mean coverage of the llvm.py file, or the diff for adding custom syntax for llvm functions? If coverage of llvm.py, then that's great but should be in a separate PR. The pytest tests don't depend on custom syntax for llvm.func

File: tests/filecheck/dialects/llvm/global.mlir
Comment (sueszli) [reply tests/filecheck/dialects/llvm/global.mlir:1]: This was intentional - just aggressively trying to maximize coverage. But it's a bit pointless.

File: tests/dialects/test_llvm.py
Comment (sueszli) [reply tests/dialects/test_llvm.py:1]: Yes - the coverage of the `llvm.py` file. I'll revert the unnecessary tests here and make a new PR

File: xdsl/dialects/llvm.py (hunk 14 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1891]: Resolved: https://github.com/xdslproject/xdsl/pull/5602

File: tests/filecheck/dialects/llvm/global.mlir
Comment (sueszli) [reply tests/filecheck/dialects/llvm/global.mlir:1]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/changes/7f17ee43737c1a9d7a7c6a9f07da2cb4e4b8d38c

File: xdsl/dialects/llvm.py (hunk 10 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py]: Not quite, can you please add an attribute dictionary with something like `{hello = "world}` and test that it's parsed and re-printed correctly?

File: tests/dialects/test_llvm.py
Comment (sueszli) [reply tests/dialects/test_llvm.py:1]: Resolved: https://github.com/xdslproject/xdsl/pull/5603

**Comment by sueszli** [comment]
Moved the pretty `llvm.return` and coverage tests into a seperate PR.

This PR should be more focused now.

Will resolve open comments within this PR now.

File: xdsl/dialects/llvm.py (hunk 10 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Thank you for the clear example - now I suddenly see what you meant!

Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/f1203763dc90492dd4cd0127f14c46ab985594f9

File: tests/filecheck/dialects/llvm/func.mlir [hunk 15]
```diff
@@ -22,28 +22,70 @@
         visibility_ = 0 : i64
     }> ({
   ^bb0(%arg0: i32, %arg1: i32):
-    llvm.return %arg0 : i32
+    "llvm.return"(%arg0) : (i32) -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/llvm/func.mlir]: Let's revert this?

File: tests/filecheck/dialects/llvm/pointers.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/llvm/pointers.mlir:1]: Also seems spurious?

File: tests/filecheck/dialects/llvm/global.mlir
Comment (superlopuh) [reply tests/filecheck/dialects/llvm/global.mlir:1]: Still here :)

File: xdsl/dialects/utils/format.py [hunk 16]
```diff
@@ -181,6 +181,28 @@ def parse_for_op_like(
...
+        printer.print_string("(")
+    if res_attrs is not None:
+        printer.print_list(
+            zip(outputs, res_attrs),
+            lambda t: print_func_output(printer, t[0], t[1]),
+        )
+    else:
+        printer.print_list(outputs, printer.print_attribute)
+    if needs_parens:
+        printer.print_string(")")
```
Comment (superlopuh) [inline xdsl/dialects/utils/format.py:217]: This PR is shaping up! I'd still like to merge things in small, semantically atomic, chunks. This seems like a great PR to merge first, with a refactor that enables functionality to be merged with only the necessary changes. Can you please take this change out into a new PR, and also extract the equivalent parsing helper? Even if they're private, it would be nice to add a doc string to describe what syntax these helpers support.

File: xdsl/dialects/llvm.py [hunk 17]
```diff
@@ -288,6 +294,21 @@ def _parse_attr_or_variadic() -> Attribute | EllipsisType:
         return [ArrayAttr(inputs), output, is_varargs]
 
 
+LINKAGE_OPTIONS = [
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: Can you please add a doc string?

File: xdsl/dialects/llvm.py [hunk 18]
```diff
@@ -1626,6 +1634,18 @@ def verify(self):
                 raise VerifyException("target features must start with '+' or '-'")
 
 
+FUNC_OP_RESERVED_ATTR_NAMES = (
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: let's make this private?

File: xdsl/dialects/llvm.py [hunk 19]
```diff
@@ -1641,6 +1661,7 @@ class FuncOp(IRDLOperation):
     # The following properties are not yet verified by the xDSL verifier, but
     # are verified to at least allow the IR to be parsed and printed correctly.
     arg_attrs = opt_prop_def(ArrayAttr[DictionaryAttr])
+    res_attrs = opt_prop_def(ArrayAttr[DictionaryAttr])
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:1677]: this functionality is not custom format, so naturally belongs in its own PR. Can you please open a separate one for it? Probably after we merge your PR that extends code coverage, at it should probably be another pytest case.

File: tests/filecheck/dialects/llvm/pointers.mlir
Comment (sueszli) [reply tests/filecheck/dialects/llvm/pointers.mlir:1]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/534ba80223c04346a5d52fbefbe35916cb758a16

File: tests/filecheck/dialects/llvm/global.mlir
Comment (sueszli) [reply tests/filecheck/dialects/llvm/global.mlir:1]: Sorry, Should be resolved now: https://github.com/xdslproject/xdsl/pull/5588/commits/534ba80223c04346a5d52fbefbe35916cb758a16

File: tests/filecheck/dialects/llvm/func.mlir (hunk 15 above)
Comment (sueszli) [reply tests/filecheck/dialects/llvm/func.mlir]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/d32c5e540434b646d11c1ca15120b9350a91beb5

File: xdsl/dialects/llvm.py (hunk 18 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/b5d5d3b60efe8c33d9bdb0b9bf2fff35cdc4a3ca

File: xdsl/dialects/llvm.py (hunk 17 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/402fc420d912f9b6b608df5f2d7c8543dbd32cb3

File: xdsl/dialects/utils/format.py (hunk 16 above)
Comment (sueszli) [reply xdsl/dialects/utils/format.py:217]: Done: #5636

**Comment by sueszli** [comment]
Set to draft. Will break down into smaller PRs.

File: xdsl/dialects/llvm.py (hunk 19 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1677]: Thank you.

Does this merged PR suffice?

https://github.com/xdslproject/xdsl/commit/a786d649ce4072b6638827202037ec48e64480ea

**Comment by sueszli** [comment]
After merging the latest changes from the main branch,
could this PR be ready for merge?

Its scope should now be limited.

File: xdsl/dialects/llvm.py (hunk 19 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py:1677]: I think this extra property deserves its own PR.

Comment (sueszli) [reply xdsl/dialects/llvm.py:1677]: Makes sense - will add.

Comment (sueszli) [reply xdsl/dialects/llvm.py:1677]: Done: https://github.com/xdslproject/xdsl/pull/5640

File: tests/filecheck/dialects/llvm/func.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/llvm/func.mlir]:
```suggestion
// RUN: XDSL_ROUNDTRIP
```

File: tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/func.mlir
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/func.mlir]:
```suggestion
// RUN: MLIR_ROUNDTRIP
// RUN: MLIR_GENERIC_ROUNDTRIP
```

Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/func.mlir]: Can you please change the input to custom format also?

**Comment by sueszli** [comment]
I noticed a mistake in my implementation regarding `unnamed_addr` and `visibility_`.

Experiment (authored by Opus 4.5): https://gist.github.com/sueszli/c33b2ef81278ab8791c97d7f91826789

Patching now.

**Comment by sueszli** [comment]
This turned out to be a very interesting edge case.

The [`func_props.mlir`](https://github.com/xdslproject/xdsl/pull/5588/changes#diff-1b57da3bf0672cfb5866be475efb9ce24247271fa8e13c48f1dbd2b95d8088af) tests cover all cases discussed in the experiments [mentioned above](https://github.com/xdslproject/xdsl/pull/5588#issuecomment-3872293982) and currently pass.

Happy to iterate further based on feedback :)

File: xdsl/dialects/llvm.py [hunk 20]
```diff
@@ -1638,7 +1669,7 @@ class FuncOp(IRDLOperation):
     CConv = prop_def(CallingConventionAttr)
     linkage = prop_def(LinkageAttr)
     sym_visibility = opt_prop_def(StringAttr)
-    visibility_ = prop_def(IntegerAttr[IntegerType])
+    visibility_ = prop_def(IntegerAttr[IntegerType], default_value=IntegerAttr(0, 64))
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:1672]: Can you please pull out all the non-custom-format related changes to FuncOp to own PRs? In this case I think the new doc string, this visibility default value, and the extra_attrs init arg. I think the three changes are small enough to go into one PR, but I'd like to iterate with you there on phrasing of the doc string and pytest for the other two things

File: xdsl/dialects/llvm.py [hunk 21]
```diff
@@ -1688,6 +1720,171 @@ def __init__(
...
+    def _print_llvm_visibility(printer: Printer, visibility: int):
+        match visibility:
+            case 0:
+                pass
+            case 1:
+                printer.print_string(" hidden")
+            case 2:
+                printer.print_string(" protected")
+            case _:
+                raise VerifyException("Invalid visibility value")
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py]: Wrong time to throw verify exceptions, we can assume that the op has been verified before printing in custom format
```suggestion
```

File: xdsl/dialects/llvm.py [hunk 22]
```diff
@@ -1688,6 +1720,171 @@ def __init__(
...
+            linkage=LinkageAttr(linkage) if linkage else LinkageAttr("external"),
+            cconv=CallingConventionAttr(cconv)
+            if cconv
+            else CallingConventionAttr("ccc"),
+            visibility=llvm_visibility,
+            sym_visibility=visibility,
+            body=region,
+            other_props=other_props,
+            extra_attrs=dict(extra_attrs.data) if extra_attrs else None,
+        )
```
Comment (superlopuh) [inline xdsl/dialects/llvm.py:1818]: Let's make the helpers return the right type so that we don't need to add more conversions here?

File: xdsl/dialects/llvm.py (hunk 21 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py]: Pyright doesn't allow this.

```
error: Cases within match statement do not exhaustively handle all values
    Unhandled type: "int"
    If exhaustive handling is not intended, add "case _: pass" (reportMatchNotExhaustive)
```

File: xdsl/dialects/llvm.py (hunk 22 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1818]: Resolved: https://github.com/xdslproject/xdsl/pull/5588/commits/40bb3aa3693dd40816f6df2b85ec50d6922d4151

File: xdsl/dialects/llvm.py (hunk 21 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py]: I would raise an AssertionError in this case

File: xdsl/dialects/llvm.py (hunk 20 above)
Comment (sueszli) [reply xdsl/dialects/llvm.py:1672]: Sounds good! Thank you for reviewing.

I've created a new PR for these 3 changes here: https://github.com/xdslproject/xdsl/pull/5647

**Comment by superlopuh** [comment]
Hi @sueszli, I think with the conflicts resolved this is good to merge! Thanks a lot for iterating here with me.

**Comment by sueszli** [comment]
Fantastic, thank you for reviewing!

**Comment by sueszli** [comment]
Next I'll work on:

- https://github.com/xdslproject/xdsl/pull/5582
- https://github.com/xdslproject/xdsl/pull/5580
- https://github.com/xdslproject/xdsl/pull/5579
- https://github.com/xdslproject/xdsl/pull/5578
- https://github.com/xdslproject/xdsl/pull/5577
- https://github.com/xdslproject/xdsl/pull/5576

**Comment by sueszli** [comment]
( ノ ^o^)ノ Yippieeeeeee
